File: android/android/app/src/main/java/com/ambitsyncmodern/hrstrap/HrStrapModule.kt
package com.ambitsyncmodern.hrstrap

import android.Manifest
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.content.ContextCompat
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.PermissionAwareActivity
import com.facebook.react.modules.core.PermissionListener
import java.util.UUID

/*
 * HR-strap R-R reader for morning HRV (COOSPO HW9 optical armband, or any strap that reports
 * R-R). SEPARATE standard-GATT peripheral, nothing to do with the watch: scan Heart Rate service
 * 0x180D -> connect -> enable Heart Rate Measurement 0x2A37 notifications via the 0x2902 CCCD ->
 * collect R-R for `seconds` -> resolve { mac, name, rrMs: [...] }. The HRV math (RMSSD etc.) is
 * done in TS (src/services/hrv.ts) so it stays identical to the desktop's tools/hrv.py.
 *
 * Modeled on AmbitSmartSensorModule (same scan/permission/GATT scaffolding); the difference is we
 * collect a timed R-R STREAM (many notifications) rather than one HR sample, and we extract the
 * RR-Interval field, not the HR. Read-only - cannot brick anything.
 */
class HrStrapModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "HrStrap"

    private val mainHandler = Handler(Looper.getMainLooper())
    private var scanCallback: ScanCallback? = null
    private var gatt: BluetoothGatt? = null
    private var pendingPromise: Promise? = null
    private var pendingPermissionPromise: Promise? = null
    private var resolved = false
    private var seconds = 120
    private var nameFilter: String? = null
    private var deviceName: String? = null
    private var deviceAddr: String? = null
    private val rrMs = ArrayList<Int>()

    companion object {
        private const val TAG = "HrStrap"
        private const val SCAN_TIMEOUT_MS = 15_000L
        private const val PERMISSION_REQUEST_CODE = 4246
        private val HR_SERVICE = uuid16("180D")
        private val HR_MEASUREMENT = uuid16("2A37")
        private val CCCD = uuid16("2902")
        private fun uuid16(short: String): UUID =
            UUID.fromString("0000$short-0000-1000-8000-00805f9b34fb")
    }

    // measure(seconds, nameFilter): collect R-R for `seconds`, matching a name substring (or any
    // HR-service peripheral when nameFilter is null/empty).
    @ReactMethod
    fun measure(seconds: Int, nameFilter: String?, promise: Promise) {
        if (pendingPromise != null) { promise.reject("BUSY", "A measurement is already running"); return }
        this.seconds = if (seconds in 20..600) seconds else 120
        this.nameFilter = nameFilter?.takeIf { it.isNotBlank() }
        if (!hasPermissions()) { pendingPermissionPromise = promise; requestPermissions(); return }
        startScan(promise)
    }

    private fun startScan(promise: Promise) {
        val adapter = (reactContext.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter
        if (adapter == null || !adapter.isEnabled) { promise.reject("BLUETOOTH_OFF", "Bluetooth is off or unavailable"); return }
        val scanner = adapter.bluetoothLeScanner
            ?: run { promise.reject("BLE_UNAVAILABLE", "BLE scanning unavailable"); return }

        pendingPromise = promise
        resolved = false
        rrMs.clear(); deviceName = null; deviceAddr = null

        val settings = ScanSettings.Builder().setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY).build()
        val timeout = Runnable {
            try { scanner.stopScan(scanCallback) } catch (_: SecurityException) {}
            rejectOnce("NOT_FOUND", "No heart-rate strap found - turn it on / wear it and retry")
        }
        mainHandler.postDelayed(timeout, SCAN_TIMEOUT_MS)

        try {
            scanner.startScan(null, settings, object : ScanCallback() {
                override fun onScanResult(callbackType: Int, result: ScanResult) {
                    val record = result.scanRecord
                    val advertisesHr = record?.serviceUuids?.any { it.uuid == HR_SERVICE } == true
                    val name = result.device.name ?: record?.deviceName
                    val filter = nameFilter
                    val nameMatches = filter != null && name != null && name.contains(filter, ignoreCase = true)
                    // Match: a name substring if given, else any peripheral advertising the HR service.
                    if (filter != null) { if (!nameMatches) return } else if (!advertisesHr) return
                    Log.d(TAG, "match: name=$name addr=${result.device.address}")
                    deviceName = name; deviceAddr = result.device.address
                    mainHandler.removeCallbacks(timeout)
                    try { scanner.stopScan(this) } catch (_: SecurityException) {}
                    connect(result.device)
                }
                override fun onScanFailed(errorCode: Int) {
                    mainHandler.removeCallbacks(timeout)
                    rejectOnce("SCAN_FAILED", "BLE scan failed, code=$errorCode")
                }
            }.also { scanCallback = it })
        } catch (e: SecurityException) {
            mainHandler.removeCallbacks(timeout)
            rejectOnce("PERMISSION_DENIED", e.message ?: "Bluetooth permission denied")
        }
    }

    private fun connect(device: BluetoothDevice) {
        try { gatt = device.connectGatt(reactContext, false, gattCallback) }
        catch (e: SecurityException) { rejectOnce("PERMISSION_DENIED", e.message ?: "connect denied") }
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                try { g.discoverServices() } catch (_: SecurityException) {}
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                // Dropped mid-capture: resolve with whatever R-R we collected.
                if (!resolved) finishSuccess()
            }
        }

        override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
            val hr = g.getService(HR_SERVICE)?.getCharacteristic(HR_MEASUREMENT)
            if (hr == null) { rejectOnce("NO_HR", "Strap has no Heart Rate Measurement characteristic"); return }
            try {
                g.setCharacteristicNotification(hr, true)
                hr.getDescriptor(CCCD)?.let {
                    it.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    g.writeDescriptor(it)
                }
                // Collect for the requested window, then resolve.
                mainHandler.postDelayed({ if (!resolved) finishSuccess() }, seconds * 1000L)
            } catch (_: SecurityException) { rejectOnce("PERMISSION_DENIED", "notify enable denied") }
        }

        override fun onCharacteristicChanged(g: BluetoothGatt, ch: BluetoothGattCharacteristic) {
            if (ch.uuid == HR_MEASUREMENT) parseRr(ch.value)
        }
    }

    // Extract R-R intervals (ms) from one 0x2A37 notification. flags bit0 HR format, bit3 energy,
    // bit4 RR present; RR are uint16 LE in 1/1024 s.
    private fun parseRr(value: ByteArray?) {
        if (value == null || value.isEmpty()) return
        val flags = value[0].toInt()
        var i = 1
        i += if (flags and 0x01 != 0) 2 else 1
        if (flags and 0x08 != 0) i += 2
        if (flags and 0x10 != 0) {
            while (i + 1 < value.size) {
                val raw = (value[i].toInt() and 0xFF) or ((value[i + 1].toInt() and 0xFF) shl 8)
                rrMs.add(Math.round(raw * 1000.0 / 1024.0).toInt())
                i += 2
            }
        }
    }

    private fun finishSuccess() {
        if (resolved) return
        resolved = true
        closeGatt()
        val arr = Arguments.createArray().apply { rrMs.forEach { pushInt(it) } }
        val map = Arguments.createMap().apply {
            putString("mac", deviceAddr ?: "")
            deviceName?.let { putString("name", it) }
            putArray("rrMs", arr)
        }
        pendingPromise?.resolve(map)
        pendingPromise = null
    }

    private fun rejectOnce(code: String, message: String?) {
        if (resolved) return
        resolved = true
        closeGatt()
        pendingPromise?.reject(code, message ?: "Error")
        pendingPromise = null
    }

    private fun closeGatt() {
        try { gatt?.disconnect(); gatt?.close() } catch (_: SecurityException) {}
        gatt = null
    }

    private fun hasPermissions(): Boolean {
        val perms = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
            arrayOf(Manifest.permission.BLUETOOTH_SCAN, Manifest.permission.BLUETOOTH_CONNECT)
        else arrayOf(Manifest.permission.ACCESS_FINE_LOCATION)
        return perms.all { ContextCompat.checkSelfPermission(reactContext, it) == PackageManager.PERMISSION_GRANTED }
    }

    private fun requestPermissions() {
        val activity = reactContext.currentActivity as? PermissionAwareActivity
        if (activity == null) {
            pendingPermissionPromise?.let { pendingPermissionPromise = null; it.reject("NO_ACTIVITY", "No active activity") }
            return
        }
        val perms = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S)
            arrayOf(Manifest.permission.BLUETOOTH_SCAN, Manifest.permission.BLUETOOTH_CONNECT)
        else arrayOf(Manifest.permission.ACCESS_FINE_LOCATION)
        activity.requestPermissions(perms, PERMISSION_REQUEST_CODE, object : PermissionListener {
            override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, results: IntArray): Boolean {
                if (requestCode != PERMISSION_REQUEST_CODE) return false
                val granted = results.isNotEmpty() && results.all { it == PackageManager.PERMISSION_GRANTED }
                val promise = pendingPermissionPromise
                pendingPermissionPromise = null
                if (granted && promise != null) startScan(promise)
                else promise?.reject("PERMISSION_DENIED", "Bluetooth permission was not granted")
                return true
            }
        })
    }
}

File: android/ios/AmbitBle/HrStrapModule.swift
//
//  HrStrapModule.swift — read R-R from a standard BLE heart-rate strap for morning HRV.
//
//  iOS twin of the Android HrStrapModule.kt and the desktop tools/hrv_strap.py. A SEPARATE
//  standard-GATT peripheral (COOSPO HW9 optical armband, or any strap that reports R-R) — NOT
//  the watch: this is a normal CoreBluetooth *central* (unlike AmbitBleModule, whose watch link
//  is an inverted GATT server). Scan Heart Rate service 0x180D → connect → subscribe Heart Rate
//  Measurement 0x2A37 → collect R-R for `seconds` → resolve { mac, name, rrMs: [Int] }.
//  The HRV math (RMSSD etc.) is the shared TS in src/services/hrv.ts, identical to the desktop.
//
import Foundation
import CoreBluetooth
import React

private let HR_SERVICE = CBUUID(string: "180D")
private let HR_MEAS = CBUUID(string: "2A37")

@objc(HrStrap)
class HrStrapModule: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {

  private var central: CBCentralManager!
  private let bleQueue = DispatchQueue(label: "com.ambitsyncmodern.hrstrap")
  private var peripheral: CBPeripheral?
  private var resolve: RCTPromiseResolveBlock?
  private var reject: RCTPromiseRejectBlock?
  private var seconds: Int = 120
  private var nameFilter: String?
  private var rrMs: [Int] = []
  private var devName: String?
  private var finished = false
  private var wantScan = false

  override init() {
    super.init()
    central = CBCentralManager(delegate: self, queue: bleQueue)
  }

  @objc static func requiresMainQueueSetup() -> Bool { false }

  // measure(seconds, nameFilter): collect R-R for `seconds`, matching a name substring, or any
  // HR-service peripheral when nameFilter is null/empty.
  @objc(measure:nameFilter:resolver:rejecter:)
  func measure(_ seconds: NSNumber,
               nameFilter: NSString?,
               resolver: @escaping RCTPromiseResolveBlock,
               rejecter: @escaping RCTPromiseRejectBlock) {
    bleQueue.async {
      if self.resolve != nil { rejecter("BUSY", "A measurement is already running", nil); return }
      let s = seconds.intValue
      self.seconds = (s >= 20 && s <= 600) ? s : 120
      let f = nameFilter as String?
      self.nameFilter = (f?.isEmpty == false) ? f : nil
      self.rrMs = []; self.devName = nil; self.peripheral = nil; self.finished = false
      self.resolve = resolver; self.reject = rejecter
      if self.central.state == .poweredOn { self.startScan() } else { self.wantScan = true }
    }
  }

  private func startScan() {
    wantScan = false
    central.scanForPeripherals(withServices: nil,
                               options: [CBCentralManagerScanOptionAllowDuplicatesKey: false])
    // Scan timeout: if nothing matched, fail cleanly.
    bleQueue.asyncAfter(deadline: .now() + 15) { [weak self] in
      guard let self = self, self.peripheral == nil, !self.finished else { return }
      self.central.stopScan()
      self.failOnce("NOT_FOUND", "No heart-rate strap found - turn it on / wear it and retry")
    }
  }

  func centralManagerDidUpdateState(_ c: CBCentralManager) {
    if c.state == .poweredOn && wantScan { startScan() }
    else if c.state != .poweredOn && wantScan { failOnce("BLUETOOTH_OFF", "Bluetooth is off") }
  }

  func centralManager(_ c: CBCentralManager, didDiscover p: CBPeripheral,
                      advertisementData: [String: Any], rssi RSSI: NSNumber) {
    let advName = (advertisementData[CBAdvertisementDataLocalNameKey] as? String) ?? p.name
    let advUuids = (advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID]) ?? []
    let advertisesHr = advUuids.contains(HR_SERVICE)
    if let filter = nameFilter {
      guard let n = advName, n.range(of: filter, options: .caseInsensitive) != nil else { return }
    } else {
      guard advertisesHr else { return }
    }
    devName = advName
    central.stopScan()
    peripheral = p
    p.delegate = self
    central.connect(p, options: nil)
  }

  func centralManager(_ c: CBCentralManager, didConnect p: CBPeripheral) {
    p.discoverServices([HR_SERVICE])
  }

  func centralManager(_ c: CBCentralManager, didFailToConnect p: CBPeripheral, error: Error?) {
    failOnce("CONNECT_FAILED", error?.localizedDescription ?? "connect failed")
  }

  func centralManager(_ c: CBCentralManager, didDisconnectPeripheral p: CBPeripheral, error: Error?) {
    // Dropped mid-capture: resolve with whatever we have.
    if !finished { finishOnce() }
  }

  func peripheral(_ p: CBPeripheral, didDiscoverServices error: Error?) {
    guard let svc = p.services?.first(where: { $0.uuid == HR_SERVICE }) else {
      failOnce("NO_HR", "Strap has no Heart Rate service"); return
    }
    p.discoverCharacteristics([HR_MEAS], for: svc)
  }

  func peripheral(_ p: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
    guard let ch = service.characteristics?.first(where: { $0.uuid == HR_MEAS }) else {
      failOnce("NO_HR", "Strap has no Heart Rate Measurement characteristic"); return
    }
    p.setNotifyValue(true, for: ch)
    // Collect for the requested window, then resolve.
    bleQueue.asyncAfter(deadline: .now() + .seconds(seconds)) { [weak self] in
      guard let self = self, !self.finished else { return }
      self.finishOnce()
    }
  }

  func peripheral(_ p: CBPeripheral, didUpdateValueFor ch: CBCharacteristic, error: Error?) {
    guard ch.uuid == HR_MEAS, let data = ch.value else { return }
    parseRr([UInt8](data))
  }

  // R-R (ms) from one 0x2A37 notification: flags bit0 HR format, bit3 energy, bit4 RR present;
  // RR are uint16 LE in 1/1024 s.
  private func parseRr(_ v: [UInt8]) {
    if v.isEmpty { return }
    let flags = Int(v[0])
    var i = 1
    i += (flags & 0x01) != 0 ? 2 : 1
    if (flags & 0x08) != 0 { i += 2 }
    if (flags & 0x10) != 0 {
      while i + 1 < v.count {
        let raw = Int(v[i]) | (Int(v[i + 1]) << 8)
        rrMs.append(Int((Double(raw) * 1000.0 / 1024.0).rounded()))
        i += 2
      }
    }
  }

  private func finishOnce() {
    if finished { return }
    finished = true
    if let p = peripheral { central.cancelPeripheralConnection(p) }
    central.stopScan()
    var out: [String: Any] = ["mac": peripheral?.identifier.uuidString ?? "", "rrMs": rrMs]
    if let n = devName { out["name"] = n }
    resolve?(out)
    resolve = nil; reject = nil
  }

  private func failOnce(_ code: String, _ message: String) {
    if finished { return }
    finished = true
    if let p = peripheral { central.cancelPeripheralConnection(p) }
    central.stopScan()
    reject?(code, message, nil)
    resolve = nil; reject = nil
  }
}
